PL-1489 update project structure (#231)
OsomePteLtd/principles: src/frontend.md

```
src/
  components/ (can be dumb and smart)
    [All components should be placed in some entity folder or shared folder]
    bankAccount/ (entity, module, .etc)
      BankAccountTable/
        [no nested directories]

>         [avoid index.ts]

Reviewer: having `index.ts` doesn't protect us from breaking modularity, but requires developers' time and bundle size

>         [avoid index.ts]

Reviewer 2: 1) I agree with you that it does not protect modularity, but for this there are our principles and review code
2) to speed up the process of writing components: there are a lot of plugins for different IDEs
3) unfortunately we have much more critical situations with sizes, and the index.js is a drop in the ocean.

PS: I want this pattern to be respected at the component level - this is important in the long run
![image](https://user-images.githubusercontent.com/11217819/217632423-5a5df334-9ec0-460c-b879-10aa1ac1809f.png)

>         [avoid index.ts]

Reviewer: so you mean if we're going to write `import { Whatever } from '../Whatever';` instead of `import { Whatever } from './Whatever';` allows us to have loose coulpling and hight cohesion? :)

>         [avoid index.ts]

Reviewer 3: @viczhuravlev want to say, that if you don't have index file in your folder, you can't understand which elements of your folder/module are internal and which external
index file it's a good place where you can define all external elements of your folder, and can be entry point for your connection this module to others
As for me such example better
```
import { Component1, Component2, Type1, helper1 } from '../modules/module_1'
import { Component1, Component2, Type1, helper1 } from '../modules/module_2'
```
than it
```
import { Component1 } from '../modules/module_1/components/Component1'
import { Component2 } from '../modules/module_1/components/Component2'
import { Type1 } from '../modules/module_1/components/types'
import { helper1 } from '../modules/module_1/components/helpers/helper1'
import { Component1 } from '../modules/module_2/components/Component1'
import { Component2 } from '../modules/module_2/components/Component2'
import { Type1 } from '../modules/module_2/components/types'
import { helper1 } from '../modules/module_2/components/helpers/helper1'
```

>         [avoid index.ts]

Reviewer: @Lecsar i guess if `Type1` and `helper1` are reused, they should be extracted out of component folder.

Exposing only external things is good, but we should focus on development speed. Usually I don't write imports manually, my IDE writes them for me, and IDE can write `import { Component1 } from '../modules/module_1/components/Component1'` instead of `import { Component1 } from '../modules/module_1'`. And I don't want to spend time on fixing it and supporting `index.ts` manually, it's just additional cost for Osome.

>         [avoid index.ts]

Reviewer 3: @Rendalf for me it's history about internal/external things of module. If we have index file i clearly understand in one place what I export and reuse, and which interfaces I can't change because it's external/public file

        [other files should be placed in queries or services]
        BankAccountTable.ts
        BankAccountTable.test.ts
        BankAccountTableBadge.ts (child component)
        BankAccountTable.styled.ts
        BankAccountTable.hook.ts (ui hook used only in the current component)

>         icon.svg (the icon used only in the current component)

Reviewer: need to mention that can be not only svg, but any static files

    shared/
      SomeNonEntityComponent/
      media/ (shared icons)
        [no nested directories]
        icon.svg
  pages/
    bankAccount/
      BankAccountList/
        [no nested directories]
        [other files should be placed in queries or services]
        [components for the page should be placed in components folder]
        BankAccountList.page.ts
        BankAccountList.page.test.ts
  queries/
    bankAccount.query.ts (query hook)
  services/
    [can be splitted into several files]
    bankAccount/ (utils, helpers, all the things that should not be put in the components folder)
      [no nested directories]
      bankAccount.service.ts
      bankAccountJournal.service.ts
      bankAccount.hook.ts (entity hook)
    [or single file]
    auditLog.service.ts

>     hook.ts (common hook)

Reviewer 4: ?
```suggestion
    shared.hook.ts (common hook)
```

```

Notes:

- **avoid adding additional folders to the project**;
- component ui hook [example](https://github.com/OsomePteLtd/agent-factory-tickets/blob/7ca56a50993592d1dbfc62a272559423e0b16303/src/components/reconciliation/shared/Basket/Basket.tsx);
- query hook [example](https://github.com/OsomePteLtd/agent/blob/main/src/queries/auditLog.query.ts);

> - entity hook [example](https://github.com/OsomePteLtd/websome/blob/main/src/hooks/ticket/useTicketNavigation.ts);

Reviewer: what does mean `entity` hook? and where should it be placed?

Author: I mean a hook used in several components in bankAccount entity

> - common hook [example](https://github.com/OsomePteLtd/websome-accounting/blob/2e48a950f2694145211dd3f3c3183c23b43e3158/src/hooks/useEventOnReady.ts);

Reviewer 5: To make those rules clearer, maybe we can add a short description for every hook type. For example:

```
Common hook:
- is not bound to a specific component;
- used in more than one component;
- doesn't make requests directly

-> goes to src/services
```

Reviewer 6: @Gashaal pls add

Author: added https://github.com/OsomePteLtd/principles/pull/231/files#diff-ece74aed489b00001b4c8172bfbde28401504cdc565c6a47551d8529d8056987R47

```
src/
  components/ (can be dumb and smart)
    [All components should be placed in some entity folder or shared folder]
    bankAccount/ (entity, module, .etc)
      BankAccountTable/
        [no nested directories]
        [avoid index.ts]

>         [other files should be placed in queries or services]

Reviewer: i guess you meant not files, but hooks here? if so, could you move closed to hooks?

Reviewer: so could you clarify it in code, not in PR comment?

Reviewer 6: let's clarify in code

        BankAccountTable.ts
        BankAccountTable.test.ts
        BankAccountTableBadge.ts (child component)
        useBankAccountTableBadge.ts (ui hook used only in the current component)
        BankAccountTable.styled.ts
        useBankAccountTable.ts (ui hook used only in the current component)
        icon.svg|png|jpeg... (the icon or any static file used only in the current component)
    shared/
      SomeNonEntityComponent/
      media/ (shared static files)
        [no nested directories]
        icon.svg|png|jpeg...
  pages/
    bankAccount/
      BankAccountList/
        [no nested directories]
        [other files should be placed in queries or services]
        [components for the page should be placed in components folder]
        BankAccountList.page.ts

>         BankAccountList.page.test.ts

Reviewer: how about `.spec` here? for integration tests

Author: added

```
src/
  components/ (can be dumb and smart)
    [All components should be placed in some entity folder or shared folder]
    bankAccount/ (entity, module, .etc)
      BankAccountTable/
        [no nested directories]
        [avoid index.ts]
        BankAccountTable.ts
        BankAccountTable.test.ts
        BankAccountTableBadge.ts (child component)
        BankAccountTable.styled.ts

>         BankAccountTable.hook.ts (ui hook used only in the current component)

Reviewer 4: hooks can be ui-less

```suggestion
        BankAccountTable.hooks.ts (ui/logic hooks used only in the current component)
```

Author: ui-less hooks should be placed in queries or services

Reviewer: if i have several large hooks, can i have several files with names like `useBankAccountTable.ts`, `useBankAccountTableWhatever.ts`?

Reviewer: @viczhuravlev if it's supposed to be used in only one component, why we should put it to another folder?

Reviewer 2: @Rendalf 
if there is one or two hooks, then I agree that it is unnecessary.
If more, it will help keep the code clean.

Reviewer: @viczhuravlev what do you mean by `keep the code clean`? could you give us some guide about clean code?

Author: > if i have several large hooks, can i have several files with names like `useBankAccountTable.ts`, `useBankAccountTableWhatever.ts`?

Of course, I wrote `BankAccountTable.hook.ts` just to emphasize is a file with hook inside.
Maybe to avoid misunderstanding better to rename to `useBankAccountTable.ts`?
And I think we can put several local hooks to component folder.

Author: > @Rendalf if there is one or two hooks, then I agree that it is unnecessary. If more, it will help keep the code clean.

I think, generally, components are not so complicated to need a lot of hooks and folder `hooks` for them. If the component has a lot of local hooks then maybe better to split it to several components.

        icon.svg (the icon used only in the current component)

>         [avoid index.ts]

Reviewer 4: `[avoid index.ts]` is duplicated

>         [other files should be placed in queries or services]

Reviewer 4: let's move all comments to the top of the directory

    shared/
      SomeNonEntityComponent/
      media/ (shared icons)
        [no nested directories]
        icon.svg
  pages/
    bankAccount/
      BankAccountList/
        [no nested directories]

> 

Reviewer 4: see no other empty lines
```suggestion
```

        BankAccountList.page.ts
        BankAccountList.page.test.ts
        [components for the page should be placed in components folder]

>         [other files should be placed in queries or services]

Reviewer 4: same
https://github.com/OsomePteLtd/principles/pull/231/files#r1100126282

  queries/

>     bankAccount.query.ts (query hook)

Reviewer 4: ```suggestion
    bankAccount.query.ts (query hooks)
```

Reviewer 5: We consider that a hook is a query hook if it uses `react-query` and `sdk`. It might also include mutations, some data normalisation etc. We should avoid hooks that include both a query and some other business logic.

  services/
    [can be splitted into several files]
    bankAccount/ (utils, helpers, all the things that should not be put in the components folder)
      [no nested directories]
      bankAccount.service.ts
      bankAccountJournal.service.ts

>       bankAccount.hook.ts (entity hook)

Reviewer 4: it's weird seeing hooks in the `services` folder
even file extension `*.hook.ts` looks strange here
what is the problem to create `hooks` top-level directory and place all `.hooks.ts` files there?

Author: The main idea is to keep a structure as simple as possible. Very often hook is a query, and in this case, the folder queries is not needed at all.

Reviewer 4: thx, got it
but according to this logic, we should have `*.query.ts` files inside `services` as well because they are also hooks

Reviewer 4: my guess `*.hook.ts` files should be placed in the `hooks` directory, `*.query.ts` - in the `queries` directory etc
or we should have only `services` directory and place all not-component files there
and differ their purpose by an extension (`hook`, `helper`, `query`, `context` etc)

Reviewer: can we just put hooks in `.service.ts` file?
it's unobvious to see smth besides `.service` file here

Author: > it's weird seeing hooks in the `services` folder even file extension `*.hook.ts` looks strange here what is the problem to create `hooks` top-level directory and place all `*.hook.ts` files there?
I wrote BankAccountTable.hook.ts just to emphasize is a file with hook inside), I will rename it

```
src/
  components/ (can be dumb and smart)
    [All components should be placed in some entity folder or shared folder]
    bankAccount/ (entity, module, .etc)
      BankAccountTable/
        [no nested directories]
        [avoid index.ts]
        [other files should be placed in queries or services]
        BankAccountTable.ts
        BankAccountTable.test.ts (unit tests)
        BankAccountTable.spec.ts (integration tests)
        BankAccountTableBadge.ts (child component)
        useBankAccountTableBadge.ts (ui hook used only in the current component)
        BankAccountTable.styled.ts

>         useBankAccountTable.ts (ui hook used only in the current component)

Reviewer 4: I guess we should briefly define what do ui/non-ui hooks mean
since it depends on where we should place hooks (components or services)

>         useBankAccountTableBadge.ts (ui hook used only in the current component)
>         BankAccountTable.styled.ts
>         useBankAccountTable.ts (ui hook used only in the current component)
>         icon.svg|png|jpeg... (the icon or any static file used only in the current component)

Reviewer 4: alphabetical order
```suggestion
        BankAccountTable.styled.ts
        icon.svg|png|jpeg... (the icon or any static file used only in the current component)
        useBankAccountTable.ts (ui hook used only in the current component)
        useBankAccountTableBadge.ts (ui hook used only in the current component)
```

    shared/

>       SomeNonEntityComponent/
>       media/ (shared static files)
>         [no nested directories]
>         icon.svg|png|jpeg...

Reviewer 4: alphabetical order
```suggestion
      media/ (shared static files)
        [no nested directories]
        icon.svg|png|jpeg...
      SomeNonEntityComponent/
```

  pages/
    bankAccount/
      BankAccountList/
        [no nested directories]
        [other files should be placed in queries or services]

>         [components for the page should be placed in components folder]

Reviewer 6: I suggest to use everywhere "directory" instead of "folder"

        BankAccountList.page.ts
        BankAccountList.page.test.ts
  queries/
    [hooks use react-query, sdk, include mutations, some data normalisation, .etc]
    bankAccount.query.ts
  services/
    [can be splitted into several files]
    bankAccount/ (utils, helpers, all the things that should not be put in the components folder)
      [no nested directories]
      bankAccount.service.ts
      bankAccountJournal.service.ts
      useBankAccount.service.ts (hooks used only in bankAccount, don't make requests directly)
    [or single file]
    auditLog.service.ts
    useEventOnReady.service.ts (common hooks not bound to a specific component, used in more than one component, don't make requests directly)
```

Notes:

> - **avoid adding additional folders to the project**;

Reviewer 6: top-level?
